### PR TITLE
Make pagination functions stable

### PR DESCRIPTION
**Pull request by cliedeman**
Make the pagination functions stable so that they can be passed into other hooks

### CODE REVIEW

File: src/usePagination.ts [hunk 1]
```diff
@@ -31,12 +31,7 @@ function usePagination<
     ReadonlyArray<$Call<ArrayKeyReturnType<TKey>>> | null,
     PaginationFunction<TOperationType['variables']>,
 ] {
-    const [data, { loadMore, hasMore, isLoading, refetchConnection }] = useOssFragment(
-        fragmentNode,
-        fragmentRef,
-    );
-
-    return [data, { loadMore, hasMore, isLoading, refetchConnection }];
+    return useOssFragment(fragmentNode, fragmentRef);
```
Comment (morrys) [inline src/usePagination.ts]: Hi @cliedeman, can you explain why this change? 
In this way usePagination is the same as useOssFragment (the `refetch` function must not be available in usePagination)

Comment (cliedeman) [reply src/usePagination.ts]: @morrys thats a mistake. I thought I could make this optimization. I will move the memo to the correct place.

**Comment by cliedeman** [comment]
Reason for changes:

I tried to create a custom hook to encapsulate the refetch logic

```
import * as React from 'react';
import { PaginationFunction, ConnectionConfig } from 'relay-hooks';
export function usePaginationState<T>(config: ConnectionConfig,
    pf: PaginationFunction
) {
    const [error, setError] = React.useState<Error | undefined>();
    const [loading, setLoading] = React.useState<boolean>(false);
    const loadMore = React.useCallback(() => {
        if (!pf.hasMore() || pf.isLoading()) {
            return;
        }
        setLoading(true);
        pf.loadMore(
            config,
            5,
            (error?: Error) => {
                setLoading(false);
                setError(error)
            },
            {}
        );
    }, [pf, setError, setLoading, config]);
    const refetch = React.useCallback((v) => {
        setLoading(true);
        pf.refetchConnection(
            config,
            5,
            (error?: Error) => {
                setLoading(false);
                setError(error);
            },
            v,
        );
    }, [pf, setLoading, setError, config]);
    return React.useMemo(() => {
        return ({
            error,
            refetch,
            loadMore,
            loading
        })
    }, [error,
        loading,
        refetch,
        loadMore])
}```

But this kept triggering re-renders in my component because the PaginationFunction was not stable

**Comment by morrys** [comment]
hi @cliedeman, can you create a test case that simulates the rerender or a minimal project where I can test it?

Do the two PR versions that you proposed solve your problem?

File: src/usePagination.ts (hunk 1 above)
Comment (morrys) [reply src/usePagination.ts]: The change made in useOssFragment seemed correct. I meant the change in usePagination which added the refetch function.

**Comment by cliedeman** [comment]
@morrys either version would solve the problem but the memo function does have some overhead so I am hesitant to add it to `useOssFragment` because that is used everywhere and most components would not benefit because they destructure the return values.

useFragment:
```
const [data] = useOssFragment(fragmentNode, fragmentRef);
```

useRefetch
```
const [data, { refetch }] = useOssFragment(fragmentNode, fragmentRef);
return [data, refetch];
```

useRefetchable
```
const [data, { refetch }] = useOssFragment(fragmentInput, fragmentRef);
```

If these cases the refetch is already a stable function so the useMemo does not add any value.

I will get around to writing a test in thew next few days

**Comment by morrys** [comment]
hi @cliedeman,
I was considering integrating your first version:

// useOssFragment.tsx
```ts
const resolverFuncion = useMemo(() => {
         return {
             refetch: resolver.refetch,
             loadMore: resolver.loadMore,
             hasMore: resolver.hasMore,
             isLoading: resolver.isLoading,
             refetchConnection: resolver.refetchConnection,
         };
}, [resolver]);

return [data, resolverFuncion];
```
Can you update your PR so that we evaluate it?

**Comment by morrys** [comment]
Hi @cliedeman,
I updated your PR with the following changes:
  * function stability test
  * fix to make functions stable
  * added react-hooks rules in eslint

I would like to release all PRs by this weekend, let me know if you can check your issue.

Thanks

**Comment by cliedeman** [comment]
Hi @morrys 

Thanks for looking into this.
This will definitely solve my use case.
I was hesitant to do it this way because I have no idea if there is a performance overhead to adding both useMemo's. 

Especially because when you do useRefetch the useOSSFragment useMemo doesn't add any benefit 

But otherwise this looks good to me

Ciaran

**Comment by morrys** [comment]
hello @cliedeman,
look at this last commit, I think this is the final version :)

**Comment by cliedeman** [comment]
Looks good to me. So the double memo is gone now?

**Comment by morrys** [comment]
yes, now we don't need the second memo because I return the resolver directly and not just the fetch functions anymore.
In fact, also the tests implemented previously have passed :)
